### ISSUE TITLE
Sealed class serializer not used for generic

### ISSUE THREAD

**Issue by rnett** [Kotlin/kotlinx.serialization#1388]
**Describe the bug**
I'm not sure exactly what is going on, but it seems that when I have a sealed class (Result) with a type parameter that is bounded by another sealed class (Failure), Failure's children won't be found.

**To Reproduce**
```kotlin
@Serializable
sealed class FailureData {
    @Serializable
    data class ServerException(val message: String = "") : FailureData()
}

@Serializable
sealed class Result<out F : FailureData> {
    @Serializable
    data class Failure<out F : FailureData>(val failure: F) : Result<F>()
}

fun main() {
    val data: Result<FailureData> = Result.Failure(FailureData.ServerException("cls"))
    val serializer = serializer<Result<FailureData>>()
    Json { }.encodeToString(serializer, data)
}
```

**Expected behavior**
It works

**Environment**
 - Kotlin version: 1.4.31
 - Library version: 1.1.0
 - Kotlin platforms: JVM IR (it happens on the old BE too)
 - Gradle version: 6.8.2

**Comment by shanshin** [comment]
As a workaround, you can try manually adding polymorphism to the sealed class using a `SerializersModule`.
```
    val module = SerializersModule {
        polymorphic(FailureData::class) {
            subclass(FailureData.ServerException.serializer())
        }
    }
``` 
and later use this module only in problematic convertions e.g.
```
val format = Json { serializersModule = module }
...
    val data: Result<FailureData> = Result.Failure(FailureData.ServerException("cls"))
    val serializer = serializer<Result<FailureData>>()
    format.encodeToString(serializer, data)
``` 
it gives JSON string `{"type":"com.example.Result.Failure","failure":{"type":"com.example.FailureData.ServerException","message":"cls"}}`

**Comment by rnett** [comment]
Unfortunately I don't have easy access to the module (and have since reworked the hierarchy for now), but that should work for anyone else.

**Comment by shanshin** [comment]
Do you have access to the serializable classes?

**Comment by rnett** [comment]
Yeah.  And using a custom module is easier than I thought, because I didn't know about `overwriteWith`, but afaik adding a polymorphic serializer only applies on deserialization anyways, and this occurs on serialization.

**Comment by shanshin** [comment]
With `SerializersModule` works serialization and deserialization
```
    val json = format.encodeToString(serializer, data)
    val instance = format.decodeFromString(serializer, json)
    println(instance)
```
prints `Failure(failure=ServerException(message=cls))`.
Try to test it for yourself.

**Comment by vsartiushkin** [comment]
The issue does not reproduce since version 2.2.0 of Kotlin and compiler plugin. 
Fixed in https://github.com/JetBrains/kotlin/commit/687a8d6e2d8e349e7960eed0b214e631a6a5e58c
